src/gsym/parser.rs

```diff
@@ -0,0 +1,323 @@
+        // and String table.
+        let end_off = off + num_addrs as usize * addr_off_size as usize;
+        if end_off > data.len() {
+            return Err(Error::new(
+                ErrorKind::InvalidData,
+                "the size of the file is smaller than expectation (address table)",
+            ));
+        }
+        let addr_tab = &data[off..end_off];
+        off = (end_off + 0x3) & !0x3;
+        let end_off = off + num_addrs as usize * 4;
+        if end_off > data.len() {
+            return Err(Error::new(
+                ErrorKind::InvalidData,
+                "the size of the file is smaller than expectation (address data offset table)",
+            ));
+        }
+        let addr_data_off_tab = &data[off..end_off];
+        let str_tab =
+            &data[strtab_offset as usize..(strtab_offset as usize + strtab_size as usize)];
+
+        Ok(GsymContext {
+            header: Header {
+                magic,
+                version,
+                addr_off_size,
+                uuid_size,
+                base_address,
+                num_addrs,
+                strtab_offset,
+                strtab_size,
+                uuid,
+            },
+            addr_tab,
+            addr_data_off_tab,
+            str_tab,
+            raw_data: data,
+        })
+    }
+
+    #[inline(always)]
+    pub fn num_addresses(&self) -> usize {
+        self.header.num_addrs as usize
+    }
+
+    /// Get the address of the an entry in the Address Table.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid index.
+    #[inline(always)]
+    pub fn addr_at(&self, idx: usize) -> u64 {
+        assert!(idx < self.header.num_addrs as usize, "invalid index");
+        let off = idx * self.header.addr_off_size as usize;
+        let mut addr = 0u64;
+        let mut shift = 0;
+        for d in &self.addr_tab[off..(off + self.header.addr_off_size as usize)] {
+            addr |= (*d as u64) << shift;
+            shift += 8;
+        }
+        addr += self.header.base_address;
+        addr
+    }
+
+    /// Get the AddressInfo of an address given by an index.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid index.
+    #[inline(always)]
+    pub fn addr_info(&self, idx: usize) -> AddressInfo {
+        assert!(idx < self.header.num_addrs as usize, "invalid index");
+        let off = idx * 4;
+        let ad_off = decode_uword(&self.addr_data_off_tab[off..]) as usize;
+        let size = decode_uword(&self.raw_data[ad_off..]);
+        let name = decode_uword(&self.raw_data[ad_off + 4..]);
+        AddressInfo {
+            size,
+            name,
+            data: &self.raw_data[ad_off + 8..],
+        }
+    }
+
+    /// Get the string at the given offset from the String Table.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid offset.
+    #[inline(always)]
```

Comment:
I think @danielocfb meant to say please "don't" use inline here? If so, I agree. We shouldn't tell the compiler to inline something unless we have a solid reason we can point to.

Reply:
I am advocating against `always`. Always conveys the **need** to inline. I can identify two possible "needs": functional correctness and a *proven* performance reason that the compiler is unable to infer otherwise. So, in the same way we do not tag each C function with `__attribute__((always_inline))`, we shouldn't do it here either, unless we have good reason to.

Yes, it's a small thing. And yet the reasons brought up in favor of this practice are not convincing at best:
https://github.com/libbpf/blazesym/pull/17#discussion_r1024611986
https://github.com/libbpf/blazesym/pull/17#discussion_r1030892059

We haven't done any profiling with and without that I am aware of, and any performance claims without that I'd call bogus and tea leaf reading. So at this point, all this annotation does is tie the compiler's metaphorical hands behind its back. And at the same time, it is very likely to cause binary bloat down the line (yes, it's not 100% forcing attribute, but it conveys a very strong desire to the compiler and to readers of the code -- both of which should be justified, in my opinion). There is literally no reason to use it here or elsewhere in the code base for correctness that has been pointed out to me.

Reply:
Thanks for clarifying. I personally would argue strongly against including `inline` at all, but IMO it's objective to say that `#[inline(always)]` is not appropriate here -- especially for rustc which has even more capacity to optimize than normal C code.

Let me put it this way: if the compiler decides to inline this when it previously wouldn't have without the hint, then you have almost certainly made the compiler do something non-optimal. At best, you get super lucky and speed things up by a couple of cycles. If that were the case, it's overwhelmingly more likely for the compiler to have inlined it in the first place. At worse, you're slowing things down (consider that the address of the function is now not located in a single place in .text, so no cache locality, instruction decode locality, etc), or possibly also causing binary bloat with no performance win.

Reply:
First of all, I agree that adding inline randomly is bad. And I have removed inlines in my local repository before discussions here.

Since there are intensive discussions here, I ran our the one and the only one benchmark `dwarf::tests::debug_info_parse_single_threaded` earlier today.

I removed `inline` marks from decode_* functions.  The results are
   - without inline                                                                                    
     - test dwarf::tests::debug_info_parse_single_threaded ... bench:   5,915,111 ns/iter (+/- 454,306)
   - with inline                                                                                       
     - test dwarf::tests::debug_info_parse_single_threaded ... bench:   5,368,025 ns/iter (+/- 501,586)

I ran the same benchmark against inlines in `debug_info` module.  The results are
   - without inline                                                                                    
     - test dwarf::tests::debug_info_parse_single_threaded ... bench:   8,520,507 ns/iter (+/- 344,442)
   - with inline                                                                                       
     - test dwarf::tests::debug_info_parse_single_threaded ... bench:   5,532,199 ns/iter (+/- 238,797)

And remove inline from both `decode_*()` and `debug_info` module.
   - without inline
     - test dwarf::tests::debug_info_parse_single_threaded ... bench:   8,613,042 ns/iter (+/- 293,465)
   - with inline
     - test dwarf::tests::debug_info_parse_single_threaded ... bench:   5,448,355 ns/iter (+/- 302,137)

I guess PGO build will do better than me.  Like you all (include me) agree that adding random inlines is bad.  But, it is true only if it is random.